File: internal/tag/tag.go
package tag

import (
	"fmt"
	"strings"

	"m4b/internal/ffmpeg"
)

type AudioTag struct {
	Title           string
	Artist          string
	Album           string
	AlbumArtist     string
	Comment         string
	Year            string
	Genre           string
	Composer        string
	Copyright       string
	Description     string
	LongDescription string
	SortName        string
	SortAlbum       string
	SortArtist      string
	Series          string
	SeriesPart      string
	CoverPath       string
}

func (t *AudioTag) ComputeSortTags() {
	if t.Series == "" || t.SeriesPart == "" {
		return
	}
	if t.SortName == "" {
		t.SortName = fmt.Sprintf("%s %s - %s", t.Series, t.SeriesPart, t.Title)
	}
	if t.SortAlbum == "" {
		t.SortAlbum = fmt.Sprintf("%s %s", t.Series, t.SeriesPart)
	}
}

// ToFFmpegFlags returns "key=value" strings for use with ffmpeg -metadata flags.
func (t AudioTag) ToFFmpegFlags() []string {
	pairs := map[string]string{
		"title":        t.Title,
		"artist":       t.Artist,
		"album":        t.Album,
		"album_artist": t.AlbumArtist,
		"comment":      t.Comment,
		"date":         t.Year,
		"genre":        t.Genre,
		"composer":     t.Composer,
		"copyright":    t.Copyright,
		"description":  t.Description,
		"synopsis":     t.LongDescription,
		"sort_name":    t.SortName,
		"sort_album":   t.SortAlbum,
		"sort_artist":  t.SortArtist,
	}
	var flags []string
	for k, v := range pairs {
		if v != "" {
			flags = append(flags, fmt.Sprintf("%s=%s", k, v))
		}
	}
	return flags
}

func FromProbeResult(r *ffmpeg.ProbeResult) AudioTag {
	tags := r.Format.Tags
	return AudioTag{
		Title:           tagVal(tags, "title"),
		Artist:          tagVal(tags, "artist"),
		Album:           tagVal(tags, "album"),
		AlbumArtist:     tagVal(tags, "album_artist"),
		Comment:         tagVal(tags, "comment"),
		Year:            tagVal(tags, "date"),
		Genre:           tagVal(tags, "genre"),
		Composer:        tagVal(tags, "composer"),
		Copyright:       tagVal(tags, "copyright"),
		Description:     tagVal(tags, "description"),
		LongDescription: tagVal(tags, "synopsis"),
		SortName:        tagVal(tags, "sort_name"),
		SortAlbum:       tagVal(tags, "sort_album"),
		SortArtist:      tagVal(tags, "sort_artist"),
	}
}

func tagVal(tags map[string]string, key string) string {
	if tags == nil {
		return ""
	}
	if v, ok := tags[key]; ok {
		return v
	}
	return tags[strings.ToUpper(key)]
}

File: src/hearworm/tag.py
from __future__ import annotations

from dataclasses import dataclass, field

from .ffmpeg import ProbeResult


@dataclass
class AudioTag:
    title: str = ""
    artist: str = ""
    album: str = ""
    album_artist: str = ""
    comment: str = ""
    year: str = ""
    genre: str = ""
    composer: str = ""
    copyright: str = ""
    description: str = ""
    long_description: str = ""
    sort_name: str = ""
    sort_album: str = ""
    sort_artist: str = ""
    series: str = ""
    series_part: str = ""
    cover_path: str = ""

    def compute_sort_tags(self) -> None:
        if not self.series or not self.series_part:
            return
        if not self.sort_name:
            self.sort_name = f"{self.series} {self.series_part} - {self.title}"
        if not self.sort_album:
            self.sort_album = f"{self.series} {self.series_part}"

    def to_ffmpeg_flags(self) -> list[str]:
        mapping = {
            "title": self.title,
            "artist": self.artist,
            "album": self.album,
            "album_artist": self.album_artist,
            "comment": self.comment,
            "date": self.year,
            "genre": self.genre,
            "composer": self.composer,
            "copyright": self.copyright,
            "description": self.description,
            "synopsis": self.long_description,
            "sort_name": self.sort_name,
            "sort_album": self.sort_album,
            "sort_artist": self.sort_artist,
        }
        return [f"{k}={v}" for k, v in mapping.items() if v]


def from_probe_result(r: ProbeResult) -> AudioTag:
    tags = r.format.tags
    return AudioTag(
        title=tags.get("title", ""),
        artist=tags.get("artist", ""),
        album=tags.get("album", ""),
        album_artist=tags.get("album_artist", ""),
        comment=tags.get("comment", ""),
        year=tags.get("date", ""),
        genre=tags.get("genre", ""),
        composer=tags.get("composer", ""),
        copyright=tags.get("copyright", ""),
        description=tags.get("description", ""),
        long_description=tags.get("synopsis", ""),
        sort_name=tags.get("sort_name", ""),
        sort_album=tags.get("sort_album", ""),
        sort_artist=tags.get("sort_artist", ""),
    )
